MoreLinq.Test/ScanByTest.cs

```diff
@@ -0,0 +1,142 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2019 Leandro F. Vieira (leandromoh). All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq.Test
+{
+    using System;
+    using NUnit.Framework;
+
+    [TestFixture]
+    public class ScanByTest
+    {
+        [Test]
+        public void ScanByIsLazy()
+        {
+            new BreakingSequence<string>().ScanBy(
+                BreakingFunc.Of<string, int>(),
+                BreakingFunc.Of<int, char>(),
+                BreakingFunc.Of<string, int, char, char>());
+        }
+
+        [Test]
+        public void ScanBy()
+        {
+            var sourceIndex = 0;
+            var source = new[]
+            {
+                "ana",
+                "beatriz",
+                "carla",
+                "bob",
+                "davi",
+                "adriano",
+                "angelo",
+                "carlos"
+            };
+
+            var uniqueKeysIndex = 0;
+            var uniqueKeys = source.Select(name => name.First())
+                                   .Distinct()
+                                   .ToArray();
+
+            var result = source.ScanBy(
+                item =>
+                {
+                    Assert.That(item, Is.EqualTo(source[sourceIndex]));
```

Comment:
This assertion within lambdas and others alike seems a little excessive because I am not sure they are helping to test anything beyond what you get from comparing against the expected result. I'd therefore keep it simple, as follows:

```c#
[Test]
public void ScanBy()
{
    var source = new[]
    {
        "ana",
        "beatriz",
        "carla",
        "bob",
        "davi",
        "adriano",
        "angelo",
        "carlos"
    };

    var result =
        from x in
            source.ScanBy(
                item => item.First(),
                key => (Element: default(string), State: -1),
                (item, key, state) => (Element: item, State: state.State + 1))
        select (x.Value.Element, x.Key, x.Value.State);

    result.AssertSequenceEqual(
        ("ana",     'a', 0),
        ("beatriz", 'b', 0),
        ("carla",   'c', 0),
        ("bob",     'b', 1),
        ("davi",    'd', 0),
        ("adriano", 'a', 1),
        ("angelo",  'a', 2),
        ("carlos",  'c', 1));
}
```